File: server/db/sessionDao.ts
import type {
  ErrorCause,
  LearningState,
  TeachingStrategy,
  VisualAidType
} from "../../shared/api.js";
import {
  boolToInt,
  dbOrDefault,
  intToBool,
  jsonText,
  parseJson,
  type Db
} from "./sqliteHelpers.js";
import type { LearningEvent, LearningSession } from "./types.js";

export function insertSession(session: LearningSession, db?: Db): void {
  dbOrDefault(db)
    .prepare(
      `INSERT INTO session (
        id, student_id, started_at, ended_at, dominant_state, summary, event_count
      ) VALUES (?, ?, ?, ?, ?, ?, ?)`
    )
    .run(
      session.id,
      session.studentId,
      session.startedAt,
      session.endedAt,
      session.dominantState,
      session.summary,
      session.eventCount
    );
}

export function getSessionById(id: string, db?: Db): LearningSession | null {
  const row = dbOrDefault(db).prepare("SELECT * FROM session WHERE id = ?").get(id) as
    | {
        id: string;
        student_id: string;
        started_at: string;
        ended_at: string | null;
        dominant_state: LearningState | null;
        summary: string | null;
        event_count: number;
      }
    | undefined;

  return row
    ? {
        id: row.id,
        studentId: row.student_id,
        startedAt: row.started_at,
        endedAt: row.ended_at,
        dominantState: row.dominant_state,
        summary: row.summary,
        eventCount: row.event_count
      }
    : null;
}

export function insertLearningEvent(event: LearningEvent, db?: Db): void {
  dbOrDefault(db)
    .prepare(
      `INSERT INTO learning_event (
        id, session_id, student_id, question_id, sequence, student_input,
        student_answer, is_correct, knowledge_point_ids, error_cause, error_detail,
        state, state_evidence, strategy, strategy_reason, care_triggered,
        visual_aid_used, tutor_response, created_at
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
    )
    .run(
      event.id,
      event.sessionId,
      event.studentId,
      event.questionId,
      event.sequence,
      event.studentInput,
      event.studentAnswer,
      boolToInt(event.isCorrect),
      jsonText(event.knowledgePointIds),
      event.errorCause,
      event.errorDetail,
      event.state,
      event.stateEvidence,
      jsonText(event.strategy),
      event.strategyReason,
      boolToInt(event.careTriggered),
      event.visualAidUsed,
      event.tutorResponse,
      event.createdAt
    );
}

export function getLearningEventById(id: string, db?: Db): LearningEvent | null {
  const row = dbOrDefault(db).prepare("SELECT * FROM learning_event WHERE id = ?").get(id) as
    | {
        id: string;
        session_id: string;
        student_id: string;
        question_id: string | null;
        sequence: number;
        student_input: string | null;
        student_answer: string | null;
        is_correct: number | null;
        knowledge_point_ids: string | null;
        error_cause: ErrorCause | null;
        error_detail: string | null;
        state: LearningState;
        state_evidence: string;
        strategy: string;
        strategy_reason: string;
        care_triggered: number;
        visual_aid_used: VisualAidType;
        tutor_response: string;
        created_at: string;
      }
    | undefined;

  return row
    ? {
        id: row.id,
        sessionId: row.session_id,
        studentId: row.student_id,
        questionId: row.question_id,
        sequence: row.sequence,
        studentInput: row.student_input,
        studentAnswer: row.student_answer,
        isCorrect: intToBool(row.is_correct),
        knowledgePointIds: parseJson<string[]>(row.knowledge_point_ids, []),
        errorCause: row.error_cause,
        errorDetail: row.error_detail,
        state: row.state,
        stateEvidence: row.state_evidence,
        strategy: parseJson<TeachingStrategy[]>(row.strategy, []),
        strategyReason: row.strategy_reason,
        careTriggered: intToBool(row.care_triggered) ?? false,
        visualAidUsed: row.visual_aid_used,
        tutorResponse: row.tutor_response,
        createdAt: row.created_at
      }
    : null;
}

File: server/dao/session_dao.py
import sqlite3

from server.dao.helpers import bool_to_int, int_to_bool, json_text, parse_json
from server.models.entities import LearningEvent, LearningSession


def insert_session(connection: sqlite3.Connection, session: LearningSession) -> None:
    connection.execute(
        """
        INSERT INTO session (id, student_id, started_at, ended_at, dominant_state, summary, event_count)
        VALUES (?, ?, ?, ?, ?, ?, ?)
        """,
        (
            session.id,
            session.student_id,
            session.started_at,
            session.ended_at,
            session.dominant_state,
            session.summary,
            session.event_count,
        ),
    )


def get_session(connection: sqlite3.Connection, session_id: str) -> LearningSession | None:
    row = connection.execute("SELECT * FROM session WHERE id = ?", (session_id,)).fetchone()
    return LearningSession.model_validate(dict(row)) if row else None


def update_session_finish(
    connection: sqlite3.Connection,
    session_id: str,
    ended_at: str,
    dominant_state: str | None,
    summary: str,
    event_count: int,
) -> None:
    connection.execute(
        """
        UPDATE session
        SET ended_at = ?, dominant_state = ?, summary = ?, event_count = ?
        WHERE id = ?
        """,
        (ended_at, dominant_state, summary, event_count, session_id),
    )


def next_event_sequence(connection: sqlite3.Connection, session_id: str) -> int:
    row = connection.execute(
        "SELECT COALESCE(MAX(sequence), 0) + 1 AS sequence FROM learning_event WHERE session_id = ?",
        (session_id,),
    ).fetchone()
    return int(row["sequence"])


def insert_learning_event(connection: sqlite3.Connection, event: LearningEvent) -> None:
    connection.execute(
        """
        INSERT INTO learning_event (
          id, session_id, student_id, question_id, sequence, student_input, student_answer,
          is_correct, knowledge_point_ids, error_cause, error_detail, state, state_evidence,
          strategy, strategy_reason, care_triggered, visual_aid_used, tutor_response, created_at
        )
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """,
        (
            event.id,
            event.session_id,
            event.student_id,
            event.question_id,
            event.sequence,
            event.student_input,
            event.student_answer,
            bool_to_int(event.is_correct),
            json_text(event.knowledge_point_ids),
            event.error_cause,
            event.error_detail,
            event.state,
            event.state_evidence,
            json_text(event.strategy),
            event.strategy_reason,
            bool_to_int(event.care_triggered) or 0,
            event.visual_aid_used,
            event.tutor_response,
            event.created_at,
        ),
    )


def _event_from_row(row: sqlite3.Row) -> LearningEvent:
    data = dict(row)
    data["is_correct"] = int_to_bool(data["is_correct"])
    data["care_triggered"] = bool(data["care_triggered"])
    data["knowledge_point_ids"] = parse_json(data.pop("knowledge_point_ids"), [])
    data["strategy"] = parse_json(data.pop("strategy"), [])
    return LearningEvent.model_validate(data)


def list_events_by_session(connection: sqlite3.Connection, session_id: str) -> list[LearningEvent]:
    rows = connection.execute(
        "SELECT * FROM learning_event WHERE session_id = ? ORDER BY sequence ASC",
        (session_id,),
    ).fetchall()
    return [_event_from_row(row) for row in rows]


def list_recent_events_by_student(
    connection: sqlite3.Connection, student_id: str, limit: int = 10
) -> list[LearningEvent]:
    rows = connection.execute(
        """
        SELECT * FROM learning_event
        WHERE student_id = ?
        ORDER BY created_at DESC
        LIMIT ?
        """,
        (student_id, limit),
    ).fetchall()
    return [_event_from_row(row) for row in rows]


def list_recent_sessions_by_student(
    connection: sqlite3.Connection, student_id: str, limit: int = 2
) -> list[LearningSession]:
    rows = connection.execute(
        """
        SELECT * FROM session
        WHERE student_id = ? AND ended_at IS NOT NULL
        ORDER BY ended_at DESC
        LIMIT ?
        """,
        (student_id, limit),
    ).fetchall()
    return [LearningSession.model_validate(dict(row)) for row in rows]
